File: 0147-insertion-sort-list/0147-insertion-sort-list.js
/**
 * Definition for singly-linked list node.
 * LeetCode에서는 ListNode가 이미 정의되어 있으므로
 * 별도로 선언할 필요가 없습니다.
 */

/**
 * Sorts a linked list using insertion sort algorithm
 *
 * @param {ListNode | null} head
 * @returns {ListNode | null}
 */
function insertionSortList(head) {
    // 빈 리스트 또는 노드가 하나인 경우
    if (head === null || head.next === null) {
        return head;
    }

    // 더미 노드
    const dummy = {
        val: head.val,
        next: head
    };

    // 이전 노드
    let previous = dummy;

    // 현재 처리할 노드
    let current = head;

    while (current !== null) {

        // 이미 정렬된 위치에 있는 경우
        if (previous.val <= current.val) {
            previous = current;
            current = current.next;
            continue;
        }

        // current가 들어갈 위치 탐색
        let insertPosition = dummy;

        while (
            insertPosition.next !== null &&
            insertPosition.next.val <= current.val
        ) {
            insertPosition = insertPosition.next;
        }

        // 다음에 처리할 노드 저장
        const nextNode = current.next;

        // current를 올바른 위치에 삽입
        current.next = insertPosition.next;
        insertPosition.next = current;

        // 기존 위치에서는 current 제거
        previous.next = nextNode;

        // 다음 노드로 이동
        current = nextNode;
    }

    // dummy 다음부터 실제 정렬된 리스트
    return dummy.next;
}

File: 0147-insertion-sort-list/0147-insertion-sort-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode insertionSortList(ListNode head) {
        // Handle edge cases: empty list or single node
        if (head == null || head.next == null) {
            return head;
        }
      
        // Create a dummy node with value smaller than any possible value
        // This serves as the head of our sorted portion
        ListNode dummy = new ListNode(Integer.MIN_VALUE);
        dummy.next = head;
      
        // Previous node in the original list and current node to be inserted
        ListNode previousNode = head;
        ListNode currentNode = head.next;
      
        // Process each node in the original list
        while (currentNode != null) {
            // If current node is already in correct position (greater than or equal to previous)
            // Simply move forward
            if (previousNode.val <= currentNode.val) {
                previousNode = currentNode;
                currentNode = currentNode.next;
                continue;
            }
          
            // Find the correct insertion position in the sorted portion
            ListNode insertPosition = dummy;
            while (insertPosition.next.val < currentNode.val) {
                insertPosition = insertPosition.next;
            }
          
            // Remove current node from its current position
            ListNode nextNode = currentNode.next;
            previousNode.next = nextNode;
          
            // Insert current node at the found position
            currentNode.next = insertPosition.next;
            insertPosition.next = currentNode;
          
            // Move to the next node to be processed
            currentNode = nextNode;
        }
      
        // Return the head of the sorted list (skip dummy node)
        return dummy.next;
    }
}
